include/deal.II/fe/fe_enriched.h

```diff
@@ -0,0 +1,545 @@
+     * As we aggregate FESystem's internal data, we only call it after
+     * the FESystem::get_data() is called.
+     */
+    InternalData ( typename FESystem<dim,spacedim>::InternalData *fesystem_data);
+
+    /**
+     * Destructor. Deletes @p fesystem_data pointer.
+     */
+    ~InternalData();
+
+    /**
+     * Gives read-access to the pointer to a @p InternalData of the @p
+     * <code>base_no</code>th base element of FESystem's data.
+     */
+    typename FiniteElement<dim,spacedim>::InternalDataBase &
+    get_fe_data (const unsigned int base_no) const;
+
+    /**
+     * Gives read-access to the pointer to an object to which into which the
+     * <code>base_no</code>th base element will write its output when calling
+     * FiniteElement::fill_fe_values() and similar functions.
+     */
+    internal::FEValues::FiniteElementRelatedData<dim,spacedim> &
+    get_fe_output_object (const unsigned int base_no) const;
+
+    /**
+     * Aggregate FESystem's internal data. It is used it every time
+     * we call FESystem's fill_fe_values() and alike.
+     */
+    typename FESystem<dim,spacedim>::InternalData *fesystem_data;
+
+    /**
+     * For each FE used in enrichment (base number <code>i</code>) and each enrichment function
+     * (base multiplicity <code>j</code>), <code>enrichment_values[i][j]</code> will be used to store
+     * possibly requested values, gradients and hessians of enrichment function <code>j</code>.
+     *
+     * The variable is made mutable as InternalData's provided to fill_fe_values and alike
+     * are const.
+     *
+     * @note We do not want to store this information in the finite element object itself,
+     * because this would mean that (i) only one FEValues object could use a finite element object at a time,
+     * and (ii) that these objects could not be used in a multithreaded context.
+     */
+    mutable std::vector<std::vector<EnrichmentValues> > enrichment;
+  };
+
+  /**
+   * For each finite element @p i used in enrichment and each enrichment function
+   * @p j associated with it (essentially its multiplicity),
+   * @p base_no_mult_local_enriched_dofs[i][j] contains the associated local DoFs
+   * on the FE_Enriched finite element.
+   */
+  std::vector<std::vector<std::vector<unsigned int> > > base_no_mult_local_enriched_dofs;
+
+  /**
+   * Enrichment functions.
+   * The size of the first vector is the same as the number of FiniteElement spaces used
+   * with enrichment. Whereas the size of the inner vector corresponds to the number
+   * of enrichment functions associated with a single FiniteElement.
+   */
+  const std::vector<std::vector<std::function<const Function<dim> *(const typename Triangulation<dim, spacedim>::cell_iterator &) > > > enrichments;
+
+  /**
+   * Auxiliary variable used to distinguish between the case when we do enrichment
+   * and when the class simply wraps another FiniteElement.
+   */
+  const bool is_enriched;
```

Comment:
> aren't there ways to compute whether the element is enriched from other member variables

yes, exactly what is done in the constructor: loop over a vector of enrichment elements and check if all of them are FE_Nothing. I don't think this should be done every single time (i.e. on every element) we need this info. Moreover, the member is `const`, so it is initialized once in constructor and there is now way to mess up with it.

> It seems to me like you're storing duplicate information here, which is always prone to the introduction of bugs later on.

in this case i think it appropriate solution.